File: src/leetcode/FourSum2.py
class Solution(object):
    def fourSumCount(self, A, B, C, D):
        """
        :type A: List[int]
        :type B: List[int]
        :type C: List[int]
        :type D: List[int]
        :rtype: int
        """
        map = {}
        count = 0
        for a in A:
            for b in B:
                sum = a + b
                if not sum in map:
                    map[sum] = 1
                else:
                    map[sum] += 1
        for c in C:
            for d in D:
                sum = -(c + d)
                if sum in map:
                    count += map[sum]
        return count
if __name__ == "__main__":
    A = [ 1, 2]
    B = [-2,-1]
    C = [-1, 2]
    D = [ 0, 2]
    solution = Solution()
    result = solution.fourSumCount(A, B, C, D)
    print result

File: src/leetcode/FourSum2.java
package leetcode;

import java.util.*;

public class FourSum2 {
	public static int fourSumCount(int[] A, int[] B, int[] C, int[] D) {
		Map<Integer,Integer> sums = new HashMap<>();
		int count = 0;
		for(int i=0; i<A.length; i++) {
			for(int j=0; j<B.length; j++) {
				int sum = A[i] + B[j];
				if(sums.containsKey(sum))
					sums.put(sum, sums.get(sum)+1);
				else
					sums.put(sum, 1);
			}
		}
		for(int i=0; i<C.length; i++) {
			for(int j=0; j<D.length; j++){
				int sum = -(C[i] + D[j]);
				if(sums.containsKey(sum))
					count += sums.get(sum);
			}
		}
		return count;
    }
	public static void main(String[] args) {
		int[] A = new int[]{1, 2};
		int[] B = new int[]{-2,-1};
		int[] C = new int[]{-1, 2};
		int[] D = new int[]{0, 2};
		int result = fourSumCount(A, B, C, D);
		System.out.println(result);
		
	}
}
